### ISSUE TITLE
Specification of `*` and '/*' in routing

### ISSUE THREAD

**Issue by usualoma** [honojs/hono#273]
I think the top-level `*` and `/*` behave a little differently than the documentation.

https://github.com/honojs/hono/releases/tag/v1.4.0
```typescript
app.get('/', (c) => {
  return c.text('foo')
}) // <--- stop process

app.use('*', middleware()) // <--- not dispached!!
```

Currently, in both TrieRouter and RegExpRouter, `*` and `/*` have higher precedence than `/`.
( On TrieRouter, `*` has the highest priority, `/*` the next highest; on RegExpRouter, `*` and `/*` have the same priority. )

```typescript
describe('star', () => {
  const router = new TrieRouter<string>()

  router.add('GET', '/', '/')
  router.add('GET', '/*', '/*')
  router.add('GET', '*', '*')

  router.add('GET', '/x', '/x')
  router.add('GET', '/x/*', '/x/*')

  it.only('top', async () => {
    const res = router.match('GET', '/')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['*', '/*', '/'])
  })

  it.only('Under a certain path', async () => {
    const res = router.match('GET', '/x')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['*', '/*', '/x', '/x/*'])
  })
})
```

```typescript
describe('star', () => {
  const router = new RegExpRouter<string>()

  router.add('GET', '/', '/')
  router.add('GET', '/*', '/*')
  router.add('GET', '*', '*')

  router.add('GET', '/x', '/x')
  router.add('GET', '/x/*', '/x/*')

  it.only('top', async () => {
    const res = router.match('GET', '/')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['*', '/*', '/']) // => failed ['/*', '*', '/']
  })

  it.only('Under a certain path', async () => {
    const res = router.match('GET', '/x')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['*', '/*', '/x', '/x/*'])
  })
})
```

The scores appear to be as follows.

at TrieRouter

```typescript
router.add('GET', '/', '/')  // score: 1.1
router.add('GET', '/*', '/*') // score: 0.2
router.add('GET', '*', '*') // score: 0.3 - 1 => Approx. -0.7
```

at RegExpRouter

```typescript
router.add('GET', '/', '/')  // score: 1.1
router.add('GET', '/*', '/*') // score: 0.2 in RegExpRouter "/*" is an alias to "*".
router.add('GET', '*', '*') // score: 0.3
```

I think the current specification for TrieRouter is acceptable to us as users, so I can make it the official specification and adapt RegExpRouter to it.
Alternatively, we can treat '/*' as an alias to '*', which would be the same as the RegExpRouter implementation.

**Comment by yusukebe** [comment]
@usualoma Hi! Thank you for your confirmation.

> I think the top-level * and /* behave a little differently than the documentation.

I've made the mistake. I'd like to write:

```ts
app.get('/api', (c) => {
  return c.text('foo')
}) // <--- stop process

app.use('/api/*', middleware()) // <--- not dispached!!
```

> Alternatively, we can treat '/' as an alias to '', which would be the same as the RegExpRouter implementation.

I think this way of treating *'/' as an alias to ''* is better.

```ts
router.add('GET', '/', '/')  // score: 1.1
router.add('GET', '/*', '/*') // score: 0.2
router.add('GET', '*', '*') // score: 0.3
```

**Comment by yusukebe** [comment]
@usualoma 

I made PR to fixed this issue #274 
Could you review it?

**Comment by yusukebe** [comment]
This PR seems to be OK. I'll merge by myself.

**Comment by usualoma** [comment]
@yusukebe Thank you!  LGTM > #274

But what should we do about this case?
https://github.com/honojs/hono/issues/262#issuecomment-1138167383

I think the following might be in order.

* `*` : Always first priority (for middleware)
* `/*` : Same as /* in other hierarchies (for fallback)

```typescript
router.add('GET', '/graphql', 'graphql')
router.add('GET', '/*', 'everything else')
router.add('GET', '*', 'apply middleware')
```

```
GET /graphql --->
'apply middleware' => 'graphql' => 'everything else'

GET /x --->
'apply middleware' => 'everything else'
```

Ah, but I wonder if this is too difficult.

Do we make the following results the same as the other tiers, and then make it a rule to "write at the beginning if you want to give priority"?

```typescript
describe('star', () => {
  const router = new TrieRouter<string>()

  router.add('GET', '/', '/')
  router.add('GET', '/*', '/*')
  router.add('GET', '*', '*')

  router.add('GET', '/x', '/x')
  router.add('GET', '/x/*', '/x/*')

  it.only('top', async () => {
    const res = router.match('GET', '/')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['/', '/*', '*'])
  })

  it.only('Under a certain path', async () => {
    const res = router.match('GET', '/x')
    expect(res).not.toBeNull()
    expect(res.handlers).toEqual(['*', '/*', '/x', '/x/*'])
  })
})
```

But I think there are times when you want to add middleware to the whole thing later, as in this case.

```typescript
router.add('GET', '/graphql', 'graphql')
router.add('GET', '/*', 'everything else')

// Lots of other routing...

router.add('GET', '*', 'apply middleware to all the request')
```

**Comment by yusukebe** [comment]
@usualoma 

In the current implementation, the following test fails.

```ts
describe('page', () => {
  const router = new TrieRouter<string>()
  router.add('GET', '/page', 'page')
  router.add('ALL', '/*', 'fallback') // or '*'
  it('GET /page', async () => {
    const res = router.match('GET', '/page')
    expect(res.handlers).toEqual(['page', 'fallback'])
  })
})
/*
    - Expected  - 1
    + Received  + 1

      Array [
    -   "page",
        "fallback",
    +   "page",
      ]
*/
```

I've noticed it's not expected. To fix it, simply remove this line.

```diff
diff --git a/src/router/trie-router/node.ts b/src/router/trie-router/node.ts
index 11dd623..6a8e43c 100644
--- a/src/router/trie-router/node.ts
+++ b/src/router/trie-router/node.ts
@@ -147,11 +147,7 @@ export class Node<T> {
       if (pattern === '*') {
         const astNode = node.children['*']
         if (astNode) {
-          let wildcard = false
-          if (astNode.methods.length) {
-            wildcard = true
-          }
-          handlerSets.push(...this.getHandlerSets(astNode, method, wildcard))
+          handlerSets.push(...this.getHandlerSets(astNode, method))
           nextNodes.push(astNode)
         }
       }
```

Then, it will be *write at the beginning if you want to give priority*. I think it's good, simple.

> But I think there are times when you want to add middleware to the whole thing later, as in this case.

In my opinion, *if you want to apply the middleware, you should write above, not later.*

But, we have one problem. This test will fail.

https://github.com/honojs/hono/blob/500fcaa9e1e2968b33b7c10c3cede5cfd892bfa0/src/hono.test.ts#L643

Same as this issue: https://github.com/honojs/hono/pull/268#issuecomment-1136724404

I wan to write `route` method as below:

```ts
 route(path: string, app?: Hono<any>): Hono<E, P> {
     this._tempPath = path
     if (app) {
       app.routes.map((r) => {
         this.addRoute(r.method, r.path, r.handler)
       })
       this._tempPath = null
     if (!app) {
       this._tempPath = path
       return this
     }

     app.routes.map((r) => {
       r.path = mergePath(path, r.path)
     })
     const routes = [...this.routes, ...app.routes]
     // this._router = new this.routerClass() <--- clear this.router another way

     routes
       .sort((a, b) => {
         return a.order - b.order
       })
       .map((r) => {
         const path = r.path
         this.addRoute(r.method, path, r.handler)
       })

     return this
   }
```

If we can clear routes in router, these problems are solved.

How about you?

**Comment by usualoma** [comment]
@yusukebe 

OK, if we go with that idea, I think we should fix the test.
> "if you want to apply the middleware, you should write above, not later".

```diff
diff --git a/src/hono.test.ts b/src/hono.test.ts
index e8f81f0..393a2fb 100644
--- a/src/hono.test.ts
+++ b/src/hono.test.ts
@@ -615,8 +615,8 @@ describe('Hono with `app.route`', () => {
       c.res.headers.append('x-custom-b', 'b')
     })
 
-    app.route('/api', api)
     app.route('/api', middleware)
+    app.route('/api', api)
 
     app.get('/foo', (c) => c.text('bar'))
 
```

Including the patches you have indicated, I think the following changes to the whole will have the expected result.

```
diff --git a/src/hono.test.ts b/src/hono.test.ts
index e8f81f0..393a2fb 100644
--- a/src/hono.test.ts
+++ b/src/hono.test.ts
@@ -615,8 +615,8 @@ describe('Hono with `app.route`', () => {
       c.res.headers.append('x-custom-b', 'b')
     })
 
-    app.route('/api', api)
     app.route('/api', middleware)
+    app.route('/api', api)
 
     app.get('/foo', (c) => c.text('bar'))
 
diff --git a/src/router/trie-router/node.test.ts b/src/router/trie-router/node.test.ts
index f2fbeb1..192e193 100644
--- a/src/router/trie-router/node.test.ts
+++ b/src/router/trie-router/node.test.ts
@@ -469,20 +469,6 @@ describe('Sort Order', () => {
     })
   })
 
-  describe('Multi match', () => {
-    const node = new Node()
-
-    node.insert('get', '/api/*', 'a') // 1.1
-    node.insert('get', '/api/entry', 'entry') // 2.2
-    node.insert('ALL', '/api/*', 'b') // 1.2
-
-    it('get /api/entry', async () => {
-      const res = node.search('get', '/api/entry')
-      expect(res).not.toBeNull()
-      expect(res.handlers).toEqual(['a', 'b', 'entry'])
-    })
-  })
-
   describe('fallback', () => {
     describe('Blog - failed', () => {
       const node = new Node()
@@ -501,14 +487,14 @@ describe('Sort Order', () => {
 describe('Multi match', () => {
   const node = new Node()
 
-  node.insert('get', '/api/*', 'a') // 1.1
+  node.insert('get', '/api/*', 'a') // 2.1 for /api/entry
   node.insert('get', '/api/entry', 'entry') // 2.2
-  node.insert('ALL', '/api/*', 'b') // 1.2
+  node.insert('ALL', '/api/*', 'b') // 2.3 for /api/entry
 
   it('get /api/entry', async () => {
     const res = node.search('get', '/api/entry')
     expect(res).not.toBeNull()
-    expect(res.handlers).toEqual(['a', 'b', 'entry'])
+    expect(res.handlers).toEqual(['a', 'entry', 'b'])
   })
 })
 
@@ -524,7 +510,7 @@ describe('star', () => {
   it('top', async () => {
     const res = node.search('get', '/')
     expect(res).not.toBeNull()
-    expect(res.handlers).toEqual(['/*', '*', '/']) // =>  failed ['*', '/*', '/']
+    expect(res.handlers).toEqual(['/', '/*', '*']) // =>  failed ['*', '/*', '/']
   })
 
   it('Under a certain path', async () => {
diff --git a/src/router/trie-router/node.ts b/src/router/trie-router/node.ts
index 11dd623..6a8e43c 100644
--- a/src/router/trie-router/node.ts
+++ b/src/router/trie-router/node.ts
@@ -147,11 +147,7 @@ export class Node<T> {
       if (pattern === '*') {
         const astNode = node.children['*']
         if (astNode) {
-          let wildcard = false
-          if (astNode.methods.length) {
-            wildcard = true
-          }
-          handlerSets.push(...this.getHandlerSets(astNode, method, wildcard))
+          handlerSets.push(...this.getHandlerSets(astNode, method))
           nextNodes.push(astNode)
         }
       }
```

**Comment by usualoma** [comment]
RegExpRouter can support this change by the following PR.
https://github.com/honojs/hono/pull/275

**Comment by usualoma** [comment]
I disagree with the following reordering, as I find the order to be jumbled and confusing, as I have described.

```typescript
     routes
       .sort((a, b) => {
         return a.order - b.order
       })
       .map((r) => {
         const path = r.path
         this.addRoute(r.method, path, r.handler)
       })
```

**Comment by yusukebe** [comment]
> I disagree with the following reordering, as I find the order to be jumbled and confusing, as I have described.

All right!!

**Comment by yusukebe** [comment]
@usualoma 

I've merged two PRs #275, #276  to master. I tried the `graghql` scenario, it works well!

```ts
import { buildSchema } from 'graphql'
import { Hono } from 'hono'
import { graphqlServer } from 'hono/graphql-server'

const app = new Hono()

const schema = buildSchema(`
type Query {
  hello: String
}
`)

const rootValue = {
  hello: () => 'Hello Hono!',
}

app.use(
  '/graphql',
  graphqlServer({
    schema,
    rootValue,
  })
)

app.get('*', (c) => {
  return c.text('fallback')
})

export default app
```

If there is no problem, I'll release v1.4.1.
